crates/runtime/src/memory.rs

```diff
@@ -86,11 +102,24 @@ pub struct MmapMemory {
     // optimize loads and stores with constant offsets.
     pre_guard_size: usize,
     offset_guard_size: usize,
+
+    // A MemFd CoW mapping that provides the initial content of this
+    // MmapMemory, if mapped.
+    //
+    // N.B.: this comes after the `mmap` field above because it must
+    // be destructed first. It puts a placeholder mapping in place on
+    // drop, then the `mmap` above completely unmaps the region.
+    memfd: Option<MemFdSlot>,
```

Comment:
Unlike the `Static` variant of memory I think we might be able to skip the storage here entirely (and also make the `instantiate` a litte faster below with fewer syscalls) since with a runtime memory all memfd really is is the initial mapping of the cow image overlaid on top of the underlying anonymous mapping. In that sense I think we can skip most of the logic of `MemFdSlot` and avoid storing redundant data.

Now that being said I think the implementation here is correct and will work well regardless. I think it'd be fine for a follow-up to clean this up a bit.

Reply:
I played with this a bit, and it's a good thought! I think though it may be safer to go through the `MemFdSlot` for heap growth specifically in case we choose to implement some other technique (such as ftruncate on a second memfd) in the future. At least if I'm reading your suggestion right, otherwise, we'd use the `MemFdSlot`'s `instantiate` logic to set up the actual CoW mapping then throw it away (without its dtor running) and rely on the implicit alignment of the `mprotect`-growth strategy between memfd and the usual dynamic memory.

Re: syscalls on instantiate, we could indeed avoid one if we give the `MemFdSlot` a "no fixed mapping" mode, but then it takes responsibility for the pre-guard too, which complicates the pooling implementation somewhat. I'll think a bit more about this though.

Reply:
One of my worries is that there's a fair bit of tricky state with handling linear memories and having that duplicated across a few locations I think is not ideal. The `MemFdSlot` tries to take over a fair bit of the management of the memory which is also tracked by the runtime `Memory`. We discussed this a bit in person as well, but to reiterate here I think it might be good to investigate a bit to figure out whether some functions below could grow an extra parameter or two instead of storing the state internally so it could be passed in by the "source of truth". 

I don't have a great idea though for whether this is possible, so I'll leave it up to your discretion about whether it ends up working out or not.